### PR TITLE
Implement Electra epoch processing 

### PR DESCRIPTION
**Pull request by EchoAlice**

### CODE REVIEW

File: .vscode/settings.json
Comment (ralexstokes) [inline .vscode/settings.json:1]: lets drop this

Comment (ralexstokes) [reply .vscode/settings.json:1]: may wanna add `.vscode` to your [global gitignore](https://git-scm.com/docs/gitignore)

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        MAX_VALIDATORS_PER_COMMITTEE,
+        SYNC_COMMITTEE_SIZE,
+        BYTES_PER_LOGS_BLOOM,
+        MAX_EXTRA_DATA_BYTES,
+        PENDING_BALANCE_DEPOSITS_LIMIT,
+        PENDING_PARTIAL_WITHDRAWALS_LIMIT,
+        PENDING_CONSOLIDATIONS_LIMIT,
+    >,
+    context: &Context,
+) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (EchoAlice) [reply ethereum-consensus/src/electra/epoch_processing.rs:37]: Followed outpt from `state_transition_across_mupltiple_forks.rs` module

**Comment by EchoAlice** [comment]
Whenever i rebase onto an updated version of `ethereum_consensus`, should i create a commit specifically for rebasing?

i realize your change within `beacon-api-client/src/types.rs` is in this commit.

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (ralexstokes) [inline .vscode/settings.json:1]: remove this from the PR

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        MAX_VALIDATORS_PER_COMMITTEE,
+        SYNC_COMMITTEE_SIZE,
+        BYTES_PER_LOGS_BLOOM,
+        MAX_EXTRA_DATA_BYTES,
+        PENDING_BALANCE_DEPOSITS_LIMIT,
+        PENDING_PARTIAL_WITHDRAWALS_LIMIT,
+        PENDING_CONSOLIDATIONS_LIMIT,
+    >,
+    context: &Context,
+) -> std::result::Result<(), Box<dyn Error>> {
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:38]: `std::result::Result` is in the [rust prelude](https://doc.rust-lang.org/std/prelude/) so you can just say `Result `here

also we want to keep the existing error type, not sure why you turned this into a trait object

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        valid_deposits.push(deposit.clone());
+    }
+    for deposit in valid_deposits.iter() {
+        increase_balance(state, deposit.index, deposit.amount);
+        processed_amount += deposit.amount;
+        next_deposit_index += 1;
+    }
+
+    state.pending_balance_deposits =
+        List::try_from(state.pending_balance_deposits.split_off(next_deposit_index)).unwrap();
```
Comment (EchoAlice) [inline ethereum-consensus/src/electra/epoch_processing.rs:117]: Try using `.copy_within()` here

Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:117]: can just do w/ `drain` for now

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    if state.pending_balance_deposits.len() == 0 {
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:119]: `is_empty`

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 6]
```diff
@@ -0,0 +1,189 @@
+use crate::electra::{
+    compute_activation_exit_epoch, get_activation_exit_churn_limit, get_active_balance,
+    get_current_epoch, increase_balance, initiate_validator_exit, is_active_validator,
+    is_eligible_for_activation, is_eligible_for_activation_queue, switch_to_compounding_validator,
+    BeaconState, Context, Error,
+};
+use ssz_rs::List;
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:7]: nitpick
```suggestion
use crate::ssz_rs::prelude::List;
```

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+use super::decrease_balance;
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:9]: generally prefer to import things in a fully-qualified manner from the crate root

```suggestion
use crate::electra::helpers::decrease_balance;
```

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        if processed_amount + deposit.amount > available_for_processing {
+            break
+        }
+        increase_balance(state, index, amount);
+        processed_amount += amount;
+        next_deposit_index += 1;
+    }
+
+    let deposits_vec: Vec<_> = state.pending_balance_deposits.drain(next_deposit_index..).collect();
+    state.pending_balance_deposits = List::try_from(deposits_vec).map_err(|err| err.1)?;
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:118]: what I originally had in mind was to flip this around
```suggestion
    let _ = state.pending_balance_deposits.drain(..next_deposit_index);
```

would be good to play around with this and make sure it does what we expect

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        }
+
+        switch_to_compounding_validator(state, target_index, context)?;
+        let active_balance = get_active_balance(state, source_index, context);
+        decrease_balance(state, source_index, active_balance);
+        increase_balance(state, target_index, active_balance);
+        next_pending_consolidation += 1;
+    }
+
+    let consolidations_vec: Vec<_> =
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:184]: same thing to try w/ `drain` here

**Review by ralexstokes — COMMENTED** [review]
great work! moving in the right direction

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            initiate_validator_exit(state, i, context)?;
+        }
+    }
+
+    let activation_epoch = compute_activation_exit_epoch(current_epoch, context);
+    for i in 0..state.validators.len() {
+        let validator = &state.validators[i];
+        if is_eligible_for_activation(state, validator) {
+            let validator = &mut state.validators[i];
+            validator.activation_epoch = activation_epoch;
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:60]: makes the line a bit more dense but you don't quite have the stutter
```suggestion
            state.validators[i].activation_epoch = activation_epoch;
```

**Review by ralexstokes — APPROVED** [review]
looks great!

left a suggestion, not required but you may like it

also, this PR is missing the updated `process_effective_balance_updates`

and something that may not be obvious in the context of the spec gen, but we also need the updated definition of `process_epoch`

im' fine to do these last two in a different  PR

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+    context: &Context,
+) {
+    // Update effective balances with hysteresis
+    let hysteresis_increment = context.effective_balance_increment / context.hysteresis_quotient;
+    let downward_threshold = hysteresis_increment * context.hysteresis_downward_multiplier;
+    let upward_threshold = hysteresis_increment * context.hysteresis_upward_multiplier;
+    for i in 0..state.validators.len() {
+        let validator = &mut state.validators[i];
+        let balance = state.balances[i];
+        let effective_balance_limit = if has_compounding_withdrawal_credential(&validator) {
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:282]:
```suggestion
        let effective_balance_limit = if has_compounding_withdrawal_credential(validator) {
```

**Review by ralexstokes — APPROVED** [review]
looks great! there is a weird thing w/ the balance constants (see comment) but otherwise it looks good, nice work

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 12]
```diff
@@ -0,0 +1,297 @@
+use crate::{
+    electra::{
+        compute_activation_exit_epoch, decrease_balance, get_current_epoch,
+        has_compounding_withdrawal_credential,
+        helpers::{
+            get_activation_exit_churn_limit, get_active_balance, initiate_validator_exit,
+            is_eligible_for_activation_queue, switch_to_compounding_validator,
+        },
+        increase_balance, is_active_validator, is_eligible_for_activation,
+        mainnet::{MAX_EFFECTIVE_BALANCE_ELECTRA, MIN_ACTIVATION_BALANCE},
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply ethereum-consensus/src/electra/epoch_processing.rs:10]: I think the preset vs config thing was a bit of a red-herring

in any case, these values (w/ the correct preset) should be supplied via the `Context`, so let's just use those

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:10]:
```suggestion
```

File: ethereum-consensus/src/electra/epoch_processing.rs [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+    let hysteresis_increment = context.effective_balance_increment / context.hysteresis_quotient;
+    let downward_threshold = hysteresis_increment * context.hysteresis_downward_multiplier;
+    let upward_threshold = hysteresis_increment * context.hysteresis_upward_multiplier;
+    for i in 0..state.validators.len() {
+        let validator = &mut state.validators[i];
+        let balance = state.balances[i];
+        let effective_balance_limit = if has_compounding_withdrawal_credential(validator) {
+            MAX_EFFECTIVE_BALANCE_ELECTRA
+        } else {
+            MIN_ACTIVATION_BALANCE
```
Comment (ralexstokes) [inline ethereum-consensus/src/electra/epoch_processing.rs:284]:
```suggestion
            context.max_effective_balance_electra
        } else {
            context.min_activation_balance
```

**Review by ralexstokes — APPROVED** [review]
amazing!

**Pull request merged**